### PR TITLE
docs: Add Rust documentation style guide

### PR DESCRIPTION
**Pull request by f-hollow**
docs: Add Rust Documentation Style Guide

PTAL

### CODE REVIEW

File: rust-doc-style-guide [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+Do NOT turn long phrases into links
+
+  > ❌ See the [Rust Reference’s section on constant evaluation](https://doc.rust-lang.org/book/ch03-01-variables-and-mutability.html) for more information on what operations can be used when declaring constants.
+
+Also take into account the following
+
+- Do not provide a link to the same location repeatedly in the same or adjacent paragraphs without a good reason, especially using different link text.
+- Do not use the same link text to refer to different locations.
+
+  > `espup` might have a section in a book and a github repo. In this case, see the [`espup`](#espup) section and [`espup` repo](https://hackmd.io/-cf0jVTqRqm9O7GM5JgLiA?both).
```
Comment (f-hollow) [inline rust-doc-style-guide:116]: Would it be better to use the link text [`espup` repo](https://hackmd.io/-cf0jVTqRqm9O7GM5JgLiA?both) or [esp-rs/espup](https://hackmd.io/-cf0jVTqRqm9O7GM5JgLiA?both) repo?

File: rust-doc-style-guide [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+
+### Recommended Terms
+
+- _SoC_ and _chip_ -- these terms are used interchangeably. Should only one be used?
+- _Product_ -- suggested as an umbrella term for Espressif chips, modules, development boards, etc. Otherwise, you need to list all product types to avoid ambiguities.
+
+### Spelling issues
+
+- _ESP-IDF_ or _esp-idf_?
+- Cargo or cargo or `cargo`?
```
Comment (f-hollow) [inline rust-doc-style-guide:252]: These are some ideas for the terms to be standardized. Need to decide on their spelling.

You will probably include more terms like this in the future.

File: rust-doc-style-guide [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+## Terminology
+
+The content here is split into two categories:
+
+- **Recommended terms** covers which terms are better to use in which circumstances
+- **Spelling issues** lists term spellings chosen by the team
+
+### Recommended Terms
+
+- _SoC_ and _chip_ -- these terms are used interchangeably. Should only one be used?
```
Comment (f-hollow) [inline rust-doc-style-guide:246]: Also, suggest agreeing on this.

File: rust-doc-style-guide [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+
+- Introduce new terms
+
+  > Each value in Rust has an _owner_.
+  > A _hash map_ allows you to associate a value with a particular key.
+
+- Emphasize important concepts or words
+
+  > When `s` comes _into_ scope, it is valid. It remains valid until it goes _out of_ scope.
+
```
Comment (f-hollow) [inline rust-doc-style-guide:214]: Some books, like [The Rust Programming Language](https://doc.rust-lang.org/book/ch02-00-guessing-game-tutorial.html#ensuring-reproducible-builds-with-the-cargolock-file), use italics for names of files and books.

Do you think the file names should be in monospace font as proposed in the Using `monospace` section?

**Review by SergioGasquez — COMMENTED** [review]
Thanks for creating the style guide! Could you make the file an MD file, or you left it without extension on purpose?

File: rust-doc-style-guide (hunk 4 above)
Comment (f-hollow) [reply rust-doc-style-guide]: Then let's stick with the monospace font option, as the readability of italics is considered worse in general.

**Comment by f-hollow** [comment]
@SergioGasquez 

Thank you for answering all my questions! I lost the file extension somewhere during rebasing...

Implemented all your comments. Does anything else need to be done?

File: rust-doc-style-guide.md [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+  - Use: Add recommended phrases
+  - Note: Add more information if needed
+
+### Recommended Terms
+
+- Cargo
+  - Avoid: cargo
+- ESP-IDF
+  - Avoid: esp-idf, esp-idf repo
+  - Use: ESP-IDF, ESP-IDF repo
```
[the remark this answers is not in the document]
Comment (f-hollow) [reply rust-doc-style-guide.md]: Updated as follows:

- ESP-IDF or esp-idf
  - Use: esp-idf when writing about the esp-idf repo
  - Use: ESP-IDF when writing about the ESP-IDF framework or ESP-IDF programming guide

**Comment by SergioGasquez** [comment]
We should also include some formatting guidelines about writing notes as we use different formats across the book.

**Comment by SergioGasquez** [comment]
From #75: We should add to the recommended terms how to refer to `std`, `no_std`. So we avoid std/Standard library/hosted environment and no_std/core/Core library/bare metal...

**Comment by f-hollow** [comment]
@SergioGasquez 

> We should also include some formatting guidelines about writing notes as we use different formats across the book.

The _Admonitions_ section lists my suggestions based on the information aqcuired through _Admonitions in other Rust books_ below. Some icons in examples are rendered differently on GitHub, so please click the links provided and see the actual notes in the books.

## Admonitions

Just because we already use ✅, ⏳, ❌ in the book, suggest following this trend and using icons such as 🚧 or 🚨.

How about this:

- Note

  > ⚠️ Note: A note covering an important point or idea. Should be used sparingly or the readers will start ignoring them.

- Attention

  > 🚨 Attention: Should be used only in very important circumstances, such as some actions can be irreversible or can be harmful to hardware, software, etc.

## Admonitions in other Rust books

### [The Rust Programming Language](https://doc.rust-lang.org/book/foreword.html)

One [note](https://doc.rust-lang.org/book/title-page.html) like this

> 🚨 **Want a more interactive learning experience? Try out a different version of the Rust Book, featuring: quizzes, highlighting, visualizations, and more**: https://rust-book.cs.brown.edu

A lot of [notes](https://doc.rust-lang.org/book/ch01-01-installation.html) like this

> Note: If you prefer not to use `rustup` for some reason, please see the Other Rust Installation Methods page for more options.

### [The Embedded Rust Book](https://docs.rust-embedded.org/book/intro/index.html)

A lot of [notes](https://docs.rust-embedded.org/book/intro/install/linux.html) like this

> **NOTE** `arm-none-eabi-gdb` is the GDB command you'll use to debug your ARM Cortex-M programs

A lot of a [note](https://docs.rust-embedded.org/book/intro/install/verify.html) like this

> **NOTE**: Old versions of openocd, including the 0.10.0 release from 2017, do not contain the new (and preferable) interface/stlink.cfg file; instead you may need to use interface/stlink-v2.cfg or interface/stlink-v2-1.cfg.

This book also has [admonitions](https://docs.rust-embedded.org/book/peripherals/singletons.html) like this

> In software engineering, the singleton pattern is a software design pattern that restricts the instantiation of a class to one object.
>
> Wikipedia: Singleton Pattern

### [The rustup book](https://rust-lang.github.io/rustup/installation/windows.html)

No admonitions found

### [The Cargo Book](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html)

A few [notes](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html) like this

> **Note**: `crates.io` does not allow bare * versions.

### [The rustc book](https://doc.rust-lang.org/nightly/rustc/targets/index.html)

A few [admonitions](https://doc.rust-lang.org/nightly/rustc/tests/index.html) like this

> ⚠️ 🚧 This option is unstable, and requires the -Z unstable-options flag. See tracking issue #82348 for more information.

A few [notes](https://doc.rust-lang.org/nightly/rustc/instrument-coverage.html) like this

> **Note**: Coverage can also be disabled on an individual function by annotating the function with the no_coverage attribute (which requires the feature flag `#![feature(no_coverage)]`).

**Comment by SergioGasquez** [comment]
I am okay with both suggestions either using :warning: and :rotating_light: or using `mdbook-admonish`.

Thanks for the research in other Rust books!

**Comment by SergioGasquez** [comment]
Also, should we include in the guidelines where to place the links? At the moment we sometimes include them at the end of the paragraph, other times at the end of the section...

**Comment by f-hollow** [comment]
@SergioGasquez @jessebraham 

> From #75: We should add to the recommended terms how to refer to `std`, `no_std`. So we avoid std/Standard library/hosted environment and no_std/core/Core library/bare metal...

[src/overview/index.md](https://github.com/esp-rs/book/pull/75/files#diff-016ba01b87f8f6e50ef7b025d5bd3b830ffacf264a3ad0bd0540c6fc28a6237e) introduces the full names of libraries and their short names (find suggested improvements below):

> There are the following approaches to using Rust on Espressif chips:
>
>- Using the `std` library, a.k.a Standard library.
>- Using the `core` library (`no_std`), a.k.a. bare metal development.

I think it will be convenient to use `std` and `no_std` in the rest of the book and mention full library names only when it is necessary. The suggested patterns when referring to `std` and `no_std` are as follows:

- write, build, or choose between  `std` / `no_std` applications
- requires use `std` for ...
- use `std` for ...
- compile in an `std` environment
- support for `no_std` requires ... (`std` support)
- using the `std` / `no_std` approach 
- `std` projects / templates / features
- `std` / `no_std` development

Also, suggest rewriting the introduction to libraries as follows:

>- Using the [Rust Standard Library](https://doc.rust-lang.org/std/) (`std`), usually referred to as the `std` development approach.
>- Using the [Rust Core Library](https://doc.rust-lang.org/core/index.html) (`core`), usually referred to as the `no_std` development approach, a.k.a. bare metal development.

**Comment by SergioGasquez** [comment]
> @SergioGasquez @jessebraham
> 
> > From #75: We should add to the recommended terms how to refer to `std`, `no_std`. So we avoid std/Standard library/hosted environment and no_std/core/Core library/bare metal...
> 
> [src/overview/index.md](https://github.com/esp-rs/book/pull/75/files#diff-016ba01b87f8f6e50ef7b025d5bd3b830ffacf264a3ad0bd0540c6fc28a6237e) introduces the full names of libraries and their short names (find suggested improvements below):
> 
> > There are the following approaches to using Rust on Espressif chips:
> > 
> > * Using the `std` library, a.k.a Standard library.
> > * Using the `core` library (`no_std`), a.k.a. bare metal development.
> 
> I think it will be convenient to use `std` and `no_std` in the rest of the book and mention full library names only when it is necessary. The suggested patterns when referring to `std` and `no_std` are as follows:
> 
> * write, build, or choose between  `std` / `no_std` applications
> * requires `std` for ...
> * use `std` for ...
> * compile in an `std` environment
> * support for `no_std` requires ... (`std` support)
> * using the `std` / `no_std` approach
> * `std` projects / templates / features
> * `std` / `no_std` development
> 
> Also, suggest rewriting the introduction to libraries as follows:
> 
> > * Using the [Rust Standard Library](https://doc.rust-lang.org/std/) (`std`), usually referred to as the `std` development approach.
> > * Using the [Rust Core Library](https://doc.rust-lang.org/core/index.html) (`core`), usually referred to as the `no_std` development approach, a.k.a. bare metal development.

Another option is using hosted environment(when using an OS) vs bare metal like in [The Rust Embedded book](https://docs.rust-embedded.org/book/intro/no-std.html). Any opinion on this @esp-rs/espressif?

**Comment by f-hollow** [comment]
> I am okay with both suggestions either using ⚠️ and 🚨 or using `mdbook-admonish`.

If there are no downsides to using extensions like this, then `mdbook-admonish` seems like a more elegant solution. I think a decision should be made from the perspective of maintenance -- whatever is more reliable and/or easier to use.

File: rust-doc-style-guide.md [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+
+  - Use `bash` syntax highlighting
+
+- Rust declarations: commands, functions, arguments, parameters, flags, variables
+- In-line command line output
+
+  > Writing a program that prints `Hello, world!`
+
+- Data types: `i8`, `u128`
+- Names of crates, traits, libraries,
```
Comment (SergioGasquez) [inline rust-doc-style-guide.md:183]: Shall we also use it for container images names? Eg: `idf-rust` in https://esp-rs.github.io/book/installation/index.html#using-containers

File: rust-doc-style-guide.md [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+This chapter lists the terms that have inconsistencies in spelling, usage, etc.
+
+If you spot other issues with terminology, please add the terms here in alphabetical order using the formatting as follows:
+
+- _Recommended term_
+  - Avoid: Add typical phrases in which this term is found
+  - Use: Add recommended phrases
+  - Note: Add more information if needed
+
+### Recommended Terms
```
Comment (SergioGasquez) [inline rust-doc-style-guide.md:247]: Maybe we should define a term for VS Code (seems like the official wording is Visual Studio Code, but VS Code is also widely used)

File: rust-doc-style-guide.md [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+- Cargo
+  - Avoid: cargo
+- ESP-IDF or esp-idf
+  - Use: esp-idf when writing about the esp-idf repo
+  - Use: ESP-IDF when writing about the ESP-IDF framework or ESP-IDF programming guide
+- _Product_
+  - Avoid: Espressif SoCs
+  - Use: Espressif products
+  - Note: In the content of running applications, suggested as an umbrella term for Espressif chips, modules, development boards, etc. Otherwise, it might be potentially ambiguous that running an app on an SoC is the same as running it on a module.
+- _SoC_
```
Comment (SergioGasquez) [inline rust-doc-style-guide.md:258]: Should SoC names such as ESP32-C3 be written in italics?

Comment (f-hollow) [reply rust-doc-style-guide.md:258]: @SergioGasquez 

Italics are usually used to highlight or attract more attention to newly introduced terms, titles of book, articles, etc. The downside is that italics are considered harder to read. In my opinion, the names of Espressif products stand out well enough thanks to "ESP" and don't require highlighting.

But I don't have any strong opinions.

Comment (SergioGasquez) [reply rust-doc-style-guide.md:258]: Italics are not used in  `ESP-IDF Programming Guide` for SoC names, so I'd say to follow that rule. We could maybe add a guideline about it, so we ensure that the guideline is followed (atm there are some places of the book that use italics for SoC names IIRC)

File: rust-doc-style-guide.md (hunk 7 above)
Comment (f-hollow) [reply rust-doc-style-guide.md:247]: Added the term

File: rust-doc-style-guide.md (hunk 6 above)
Comment (f-hollow) [reply rust-doc-style-guide.md:190]: Added the line

- Images and containers: `idf-rust`

**Comment by f-hollow** [comment]
> > I am okay with both suggestions either using ⚠️ and 🚨 or using `mdbook-admonish`.
> 
> If there are no downsides to using extensions like this, then `mdbook-admonish` seems like a more elegant solution. I think a decision should be made from the perspective of maintenance -- whatever is more reliable and/or easier to use.

@SergioGasquez 

Added the formatting guidelines for admonitions to the style guide. The structures "⚠️ Note:" and "🚨 Warning:" render in `mdbook` perfectly. However, when you view the rendered markdown file on GitHub, you see additional icons in front of Note and Warning added by GitHub automatically.

If it doesn't bother you, let's leave it like this. Otherwise, we can delete our icons.

**Comment by f-hollow** [comment]
@SergioGasquez 

Seems like I resolved all your comments. PTAL.

File: rust-doc-style-guide.md [hunk 9]
```diff
@@ -0,0 +1,320 @@
...
+  - Use only if necessary: bare metal development, hosted environment, Standard library, Core library
+  - Use `std` / `no_std` by default in patterns, such as:
+    - write, build, or choose between  `std` / `no_std` applications
+    - requires `std` for ...
+    - use `std` for ...
+    - compile in an `std` environment
+    - support for `no_std` requires ... (`std` support)
+    - using the `std` / `no_std` approach
+    - `std` projects / templates / features
+    - `std` / `no_std` development
```
Comment (SergioGasquez) [inline rust-doc-style-guide.md:284]: We've decided to proceed with `std`/`no_std` but using bare metal for `no_std` can also be used, so I think this paragraph is pretty accurate

**Review by SergioGasquez — APPROVED** [review]
LGTM! Thanks for all your work on improving the Rust documentation!

**Review by MabezDev — APPROVED** [review]
Looks good to me :)

**Pull request merged**